Sr 11159 space engine

AndrewLitteken opened the pull request:
<!-- What's in this pull request? -->
Split off of #26741 to separate Enum changes and Space Engine changes, related to #26741

Changes Identifiers to DeclNames so that in checking elements the associated values are taken into account.  Other exhaustiveness and redundancy to compare these associated values are taken to make sure the correct message is shown. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-11159](https://bugs.swift.org/browse/SR-11159).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

CodaFi commented:
Please let's just restrict this to the SpaceEngine changes.  The engine should pick up DeclNames by projection, so the rest of this patch isn't needed.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -604,8 +612,9 @@ namespace {
             // This is especially helpful when dealing with `unknown` case
             // in parameter positions.
-            if (s1 == reducedSpace)
+            if (s1 == reducedSpace){
```
I appreciate the bracing, but please standardize on the spacing in the rest of this file.  If in doubt, run `clang-format`.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -696,9 +705,9 @@ namespace {
           break;
         case SpaceKind::Constructor: {
-          if (!Head.empty()) {
+          if (!Head.getBaseIdentifier().empty()) {
```
We're going to need to print the labels into the pattern as well.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -1435,6 +1443,11 @@ namespace {
         if (!SP) {
           // If there's no sub-pattern then there's no further recursive
           // structure here.  Yield the constructor space.
+          if(VP->hasUnresolvedOriginalExpr()){
```
The space engine does not need to concern itself with ill-formed pattern trees.  They will have been structurally diagnosed by the rest of pattern type checking, and we should not project them.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -333,7 +334,7 @@ namespace {
           }
 
-          return true;
+          return false;
```
I'm curious if this affects any of our tests.  I suspect it may close some SRs if this turns out to be salient.

AndrewLitteken replied:
I ran the `--test` with the build script, the tests still passed in those cases.

CodaFi replied:
Hm, let's leave it in and run the broader test suite to see if this catches anything.

AndrewLitteken replied:
Is that the validation-test?

CodaFi reviewed (changes requested):
The changes and the general approach look pretty good.  Some feedback inline.

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -115,7 +115,7 @@ namespace {
       // In type space, we reuse HEAD to help us print meaningful name, e.g.,
       // tuple element name in fixits.
-      Identifier Head;
+      DeclName Head;
```
This isn't your bug, but "Head" no longer makes sense under these changes.  Perhaps `Name` or `Spine` (definitely rooting more for the former than the latter).

CodaFi commented:
Let's also try to clean up the history a little here.  In general, you can run

```bash
git pull --rebase upstream master
```

where `upstream` points to the Swift remote endpoint.  This will keep the merge commits out of your history.  An interactive rebase can also help to squash away these intermediate commits.  If this is too much trouble, I can also manually squash these in after we run CI.

AndrewLitteken committed:
adding space engine changes

adding full spaceengine changes

using clang-format

adding argument name size comparison

adding space engine changes

adding full spaceengine changes

using clang-format

adding argument printing to interleave function

updating tests for new print formatting

removing extra checks from constructor subspace checking

adding alternate path if the argument and space lenght are the same

clang format

adding FIXME for code performance

adding fixme for showing constructor

switching direction of beginningparenthesis

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -372,10 +371,17 @@ namespace {
+                other.Head.getBaseIdentifier())
+              return false;
+            if (this->Head.getArgumentNames().size() != 0 &&
+                other.Head.getArgumentNames().size() != 0) {
```
This sequence of checks doesn't make sense to me.  To preserve the optimization, we should be checking if the `baseIdentifier`s are different and the `argumentNames` differ in number.

AndrewLitteken replied:
I may have misinterpreted some fo the tests.  I think this was attempting to take care of the case where if the case statement had no arguments/associated values (i.e. `.a`) compared to one where it (`.a(x:)`) then it should be considered a subspace.

CodaFi replied:
We pre-expand patterns via lookup to make that impossible.  Otherwise we wouldn't be able to detect overlapping patterns.

AndrewLitteken replied:
Ah ok, makes sense

AndrewLitteken commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -372,10 +372,17 @@ namespace {
+            if (this->Head.getArgumentNames().size() != 
+                other.Head.getArgumentNames().size()) {
+              return false;
+            }
```
should be fixed and passes the basic/validation tests from the build script

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -696,19 +698,38 @@ namespace {
+              labelSpaces.push_back(std::pair<Identifier, Space>(*args, param));
+              args++;
+            } else
+              labelSpaces.push_back(
```
Could you assert that you always have the same number of argument names in the base identifier as there are subspaces?  If that much is true, then you can swap the interleave for the variant of `std::transform` that takes two input iterators and keep track of placing the commas yourself.  Honestly, if it's all the same we can just take this code and you can leave a FIXME behind to come clean this up.  I'm not too concerned about printing being efficient since this will only come up when you're diagnosing invalid code.

AndrewLitteken replied:
Do you mean a full assertion that stops compilation? Or just a check that runs a different path if that is the case?

theblixguy replied:
Yeah, just an `assert`.

AndrewLitteken replied:
Is that really what we want to happen though? Because there could be a case where there are no argument names but there are spaces, so we wouldn't want it to stop in that case.

AndrewLitteken replied:
I am not sure just having an assertion allows all cases we would want.

For instance, if we check the number of argument names against spaces in the following example:
```enum Threepeat {
  case a, b, c
}

func test3(x: Threepeat, y: Threepeat) {
  switch (x, y) { // expected-error {{switch must be exhaustive}}
  // expected-note@-1 {{add missing case: '(.a, .c)'}}
  case (.a, .a):
    ()
  case (.b, _):
    ()
  case (.c, _):
    ()
  case (_, .b):
    ()
  }
}```
with the assertion, it fails where it should not.

AndrewLitteken commented:
I added a FIXME to the initial implementation, I'm not sure that an assertion can cover all the cases needed.

CodaFi commented:
Looks good.

@swift-ci please test

AndrewLitteken commented:
Apologies for that bug, the parenthesis should now be in the right direction, which I believe was the cause of the failures

theblixguy commented:
@swift-ci please test

CodaFi commented:
Thanks

⛵️
